disttask/framework/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,412 @@
+		logutil.BgLogger().Info("detection, task is finished", zap.Int64("taskID", gTask.ID))
+		d.delRunningGlobalTasks(gTask.ID)
+	}
+}
+
+func (d *dispatcher) updateTaskRevertInfo(gTask *proto.Task) {
+	// TODO: Add error msg to task.
+	gTask.State = proto.TaskStateReverted
+	// Write the global task meta into the storage.
+	err := d.gTaskMgr.UpdateTask(gTask)
+	if err != nil {
+		logutil.BgLogger().Warn("update global task failed", zap.Error(err))
+	}
+}
+
+func (d *dispatcher) processErrFlow(gTask *proto.Task, receiveErr string) error {
+	meta, err := GetTaskFlowHandle(gTask.Type).ProcessErrFlow(d, gTask, receiveErr)
+	if err != nil {
+		logutil.BgLogger().Warn("handle error failed", zap.Error(err))
+		return err
+	}
+
+	// TODO: Consider using a new context.
+	instanceIDs, err := d.getTaskAllInstances(d.ctx, gTask.ID)
+	if err != nil {
+		logutil.BgLogger().Warn("get global task's all instances failed", zap.Error(err))
+		return err
+	}
+
+	if len(instanceIDs) == 0 {
+		gTask.State = proto.TaskStateReverted
+	} else {
+		gTask.State = proto.TaskStateReverting
+	}
+	// Write the global task meta into the storage.
+	err = d.gTaskMgr.UpdateTask(gTask)
+	if err != nil {
+		logutil.BgLogger().Warn("update global task failed", zap.Error(err))
+		return err
+	}
+
+	// New rollback subtasks and write into the storage.
+	for _, id := range instanceIDs {
+		subtask := proto.NewSubtask(gTask.ID, gTask.Type, id, meta)
+		err = d.subTaskMgr.AddNewTask(gTask.ID, subtask.SchedulerID, subtask.Meta, subtask.Type, true)
+		if err != nil {
+			logutil.BgLogger().Warn("add subtask failed", zap.Int64("gTask ID", gTask.ID), zap.Error(err))
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *dispatcher) processNormalFlow(gTask *proto.Task, fromPending bool) (err error) {
+	// Generate the needed global task meta and subTask meta.
+	handle := GetTaskFlowHandle(gTask.Type)
+	if handle == nil {
+		logutil.BgLogger().Warn("gen gTask flow handle failed", zap.Int64("ID", gTask.ID), zap.Bool("pending", fromPending))
+		d.updateTaskRevertInfo(gTask)
+		return errors.Errorf("%s type handle doesn't register", gTask.Type)
+	}
+	finished, metas, err := handle.ProcessNormalFlow(d, gTask, fromPending)
+	if err != nil {
+		logutil.BgLogger().Warn("gen dist-plan failed", zap.Error(err))
+		return err
+	}
+
+	// Adjust global task meta.
+	if gTask.Concurrency == 0 {
+		gTask.Concurrency = DefaultSubtaskConcurrency
+	}
+	if gTask.Concurrency > MaxSubtaskConcurrency {
+		gTask.Concurrency = MaxSubtaskConcurrency
+	}
+	if finished {
+		gTask.State = proto.TaskStateSucceed
+	}
+
+	// Special handling for the new tasks.
+	// About TaskStateSucceed gTask, we consider keeping records first.
+	if gTask.State == proto.TaskStatePending {
+		// TODO: Consider using TS.
+		gTask.StartTime = time.Now().UTC()
+		gTask.State = proto.TaskStateRunning
+	} else {
+		gTask.StateUpdateTime = time.Now().UTC()
```

Comment:
When a task has serval steps, the `StateUpdateTime` also updates while step change, but the `State` does not change (Running => Running), it is a little strange

Reply:
The next step of step1-running should be step1-finish or reverting